text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,634 @@
+in order to return the correct data.
+
+```
+client -> service A -> service C
+```
+
+In this example, we would like `service A` to decide on which backend service 
+to call, based on the client version. We would also like to trace the entire 
+system, in order to understand if requests to `service C` are slower or faster 
+than `service B`. What might `service A` look like?
+
+## Global initialization 
+First, during program initialization, `service A` configures correlation and tracing 
+propagation, and include them in the global list of injectors and extractors. 
+Let's assume this tracing system is configured to use B3, and has a specific 
+propagator for that format. Initializing the propagators might look like this:
+
+```php
+func InitializeOpentelemetry() {
+  // create the propagators for tracing and correlations.
+  bagExtract, bagInject = Correlations::HTTPPropagator()
+  traceExtract, traceInject = Tracer::B3Propagator()
+  
+  // add the propagators to the global list.
+  Propagation::SetExtractors(bagExtract, traceExtract)
+  Propagation::SetInjectors(bagInject, traceInject)
+}
+```
+
+## Extracting and injecting from HTTP headers
+These propagators can then be used in the request handler for `service A`. The 
+tracing and correlations concerns use the context object to handle state without 
+breaking the encapsulation of the functions they are embedded in.
+
+```php
+func ServeRequest(context, request, project) -> (context) {
+  // Extract the context from the HTTP headers. Because the list of 
+  // extractors includes a trace extractor and a correlations extractor, the 
+  // contents for both systems are included in the  request headers into the 
+  // returned context.
+  extractors = Propagation::GetExtractors()
+  context = Propagation::Extract(context, extractors, request.Headers)
+
+  // Start a span, setting the parent to the span context received from 
+  // the client process. The new span will then be in the returned context.
+  context = Tracer::StartSpan(context, [span options])
+  
+  // Determine the version of the client, in order to handle the data 
+  // migration and allow new clients access to a data source that older 
+  // clients are unaware of.
+  version = Correlations::GetCorrelation( context, "client-version")
+
+  switch( version ){
+    case "v1.0":
+      data, context = FetchDataFromServiceB(context)
+    case "v2.0":
+      data, context = FetchDataFromServiceC(context)
+  }
+
+  context = request.Response(context, data)
+
+  // End the current span
+  Tracer::EndSpan(context)
+
+  return context
+}
+
+func FetchDataFromServiceB(context) -> (context, data) {
+  request = NewRequest([request options])
+  
+  // Inject the contexts to be propagated. Note that there is no direct 
+  // reference to tracing or correlations.
+  injectors = Propagation::GetInjectors()
+  context = Propagation::Inject(context, injectors, request.Headers)
```

Comment:
Does inject return a context or headers?

Reply:
Good catch, it returns the new headers.